tests/nn/misc/test_checkpoint_activations.py

```diff
@@ -252,3 +252,30 @@ def train(model, in1, in2):
         if no_cpt[key] != cpt[key]:
             print(no_cpt, cpt)
             assert 0
+
+
+def test_list_input():
+    class Model(nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.conv = nn.Linear(2, 2)
+
+        def forward1(self, x):
+            x[0] = self.conv(x[0])
+            return x
+
+        def forward(self, x):
+            if not isinstance(x, (list, tuple)):
+                x = [x]
```

Comment:
This is very interesting! The difference in behavior is between x and [x]. Is that right? Do the `requires_grad` field of the loss tensor change from "False" to "True" when the type() of x changes?

For OffloadModel, I had an issue with triggering the backward hook for int inputs that are common for sequence indices in Transformer models. i use a [dummy input ](https://github.com/facebookresearch/fairscale/blob/master/fairscale/experimental/nn/offload.py#L490) which means I always trigger the backward pass no matter what. Sometimes it might be wasteful if there are no grads to propagate but it handles edge cases and does not require the user to modify the input.

Reply:
Wow, the dummy is very good idea! It seems to work for the test case. Thank you so much @anj-s! Let me test with the real workload.

Reply:
Great! However I am still not sure why the error shows up? From my experience as long as one of the inputs to `forward` is a tensor with requires_grad=True it should work. I wonder why using list causes it to fail.

Reply:
yeah, it is more than that. I tested pytorch's checkpoint as well. There, the same issue exists. Worse, when I change the input from list to a tensor and use a list to wrap the tensor inside the forward function, it still doesn't work. I think the dummy tensor method works very well since we don't want to rely on user's input types to get the backward pass called. When we checkpoint, we know we want it called. Also, in the backward function, we do check for all outputs have requires_grad when running under grad_enabled. Therefore, we are pretty safe in detecting unnecessary checkpoint calls.